src/mlpack/tests/linear_svm_test.cpp

```diff
@@ -0,0 +1,1129 @@
+  BOOST_REQUIRE_EQUAL(svmfBigReg.NumFunctions(), points);
+
+
+  for (size_t i = 0; i < trials; ++i)
+  {
+    // Create a random set of parameters.
+    arma::mat parameters;
+    parameters.randu(numClasses, inputSize);
+
+    double wL2SquaredNorm;
+    wL2SquaredNorm = 0.5 * arma::accu(parameters % parameters);
+
+    // Calculate regularization terms.
+    const double smallRegTerm = 0.5 * wL2SquaredNorm;
+    const double bigRegTerm = 20 * wL2SquaredNorm;
+
+    for (size_t j = 0; j < points; ++j)
+    {
+      BOOST_REQUIRE_CLOSE(svmfNoReg.Evaluate(parameters, j, 1) + smallRegTerm,
+          svmfSmallReg.Evaluate(parameters, j, 1), 1e-5);
+      BOOST_REQUIRE_CLOSE(svmfNoReg.Evaluate(parameters, j, 1) + bigRegTerm,
+          svmfBigReg.Evaluate(parameters, j, 1), 1e-5);
+    }
+  }
+}
+
+/**
+ * Test gradient for the LinearSVMFunction.
+ */
+BOOST_AUTO_TEST_CASE(LinearSVMFunctionGradient)
+{
+  const size_t points = 500;
+  const size_t inputSize = 10;
+  const size_t numClasses = 3;
+
+  // Initialize a random dataset.
+  arma::mat data;
+  data.randu(inputSize, points);
+
+  // Create random class labels.
+  arma::Row<size_t> labels(points);
+  for (size_t i = 0; i < points; i++)
+    labels(i) = math::RandInt(0, numClasses);
+
+  // 2 objects for 2 terms in the cost function. Each term contributes towards
+  // the gradient and thus need to be checked independently.
+  LinearSVMFunction<arma::mat> svmf1(data, labels, numClasses, 0.0);
+  LinearSVMFunction<arma::mat> svmf2(data, labels, numClasses, 10.0);
+
+  // Create a random set of parameters.
+  arma::mat parameters;
+  parameters.randu(numClasses, inputSize);
+
+  // Get gradients for the current parameters.
+  arma::mat gradient1, gradient2;
+  svmf1.Gradient(parameters, gradient1);
+  svmf2.Gradient(parameters, gradient2);
+
+  // Perturbation constant.
+  const double epsilon = 0.001;
+  double costPlus1, costMinus1, numGradient1;
+  double costPlus2, costMinus2, numGradient2;
+
+
+  // For each parameter.
+  for (size_t i = 0; i < numClasses; i++)
+  {
+    for (size_t j = 0; j < inputSize; j++)
+    {
+      // Perturb parameter with a positive constant and get costs.
+      parameters(i, j) += epsilon;
+      costPlus1 = svmf1.Evaluate(parameters);
+      costPlus2 = svmf2.Evaluate(parameters);
+
+      // Perturb parameter with a negative constant and get costs.
+      parameters(i, j) -= 2 * epsilon;
+      costMinus1 = svmf1.Evaluate(parameters);
+      costMinus2 = svmf2.Evaluate(parameters);
+
+      // Compute numerical gradients using the costs calculated above.
+      numGradient1 = (costPlus1 - costMinus1) / (2 * epsilon);
+      numGradient2 = (costPlus2 - costMinus2) / (2 * epsilon);
+
+      // Restore the parameter value.
+      parameters(i, j) += epsilon;
+
+      // Compare numerical and backpropagation gradient values.
+      BOOST_REQUIRE_SMALL(numGradient1 - gradient1(i, j), 1e-2);
+      BOOST_REQUIRE_SMALL(numGradient2 - gradient2(i, j), 1e-2);
```

Comment:
Ah, I see you already have a numeric gradient test. :+1:  This is nice, but do you think we can reduce the tolerance to something more like `1e-3` or `1e-4`?

Reply:
I've been using `BOOST_REQUIRE_CLOSE` and `BOOST_REQUIRE_SMALL`with several values for tolerance including `1e-5`,`1e-4`,`1e-3`,`1e-2` and they are listed in decreasing order of the test failure (ultimately resulting in build failures). For the tolerance = `1e-2`, the build passes. 
(I think the tests on the commit history got removed since I had to rebase this with `master` after #1713). 
I will re-evaluate this but do you have a strategy that might help?

Reply:
Removed this test case as of now, and replaced it with a casual test that compares the value evaluated without matrix operations to that calculated with `Gradient()`. If I am able to fix this, I'll include this test.